soc: silabs: Initialize oscillators and clock tree from DT, clean up init sequence

asmellby opened the pull request:
This PR cleans up the Series 2 init sequence by
* Unifying all Series 2 devices on using the device init HAL (adding EFR32MG21, all the others already used it)
  * DCDC is added to DeviceTree to be able to only initialize it if present
* Migrating oscillator init from device init HAL to clock manager HAL based on DT
* Migrating clock tree init from ad hoc board and driver config to clock manager HAL based on DT
* Separating soc.c for Series 2 from Series 0 and 1, since the init sequence is fully disjoint

asmellby committed:
soc: silabs: Initialize DCDC from device tree

The DC-DC converter was unconditionally initialized with default
settings on Series 2. Add device tree binding and nodes, and guard
call to init function. Map DT options to config header from HAL.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
soc: silabs: Enable device init on EFR32MG21

Switch EFR32MG21 to use the device init HAL. This makes the init sequence
the same as the rest of Series 2.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
dts: silabs: Add clock bindings and clock tree

Introduce bindings for Series 2 oscillators: HFRCODPLL, HFRCOEM23,
LFRCO and LFXO.

Add clock tree representation in devicetree `clocks` node.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
soc: silabs: Initialize clock manager HAL from DT

Swap from the deprecated device_init_* functions to clock manager
for clock tree configuration. Populate config headers using
device tree representation of clock tree and oscillator config.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
boards: silabs: Remove clock configuration

Clock setup is now done by the clock manager based
on device tree configuration.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
drivers: timer: gecko: Remove clock configuration

Clock setup is now done by the clock manager based
on device tree configuration.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
soc: silabs: Separate Series 2 soc.c

Series 2 always uses the device init HAL, while Series 0/1 never do.
Create a separate soc.c for Series 2 to make both versions easier to read.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

jerome-pouiller commented on boards/silabs/dev_kits/sltb010a/board.c:
So good!

jerome-pouiller commented on soc/silabs/silabs_s2/soc.c:
```diff
@@ -0,0 +1,41 @@
+#ifdef CONFIG_PM
+#include <sl_hfxo_manager.h>
+#include <sl_power_manager.h>
+#endif
```
Probably it does not hurt to include `sl_hfxo_manager.h` and `sl_power_manager.h` even if `!CONFIG_PM`.

jerome-pouiller commented on soc/silabs/silabs_s2/soc.c:
```diff
@@ -0,0 +1,41 @@
+#ifdef CONFIG_PM
+	sl_power_manager_init();
+	sl_hfxo_manager_init();
+#endif
```
If you include `sl_hfxo_manager.h` and `sl_power_manager.h` unconditionally, you can change the `#ifdef CONFIG_PM` in `if (IS_ENABLED(CONFIG_PM))` and make the code easier.

jerome-pouiller commented on soc/silabs/silabs_s2/soc.c:
```diff
@@ -0,0 +1,41 @@
+
+#if CONFIG_HW_HAS_SILABS_DCDC
+	sl_device_init_dcdc();
+#endif
```
I suggest:
```
    if (IS_ENABLED(CONFIG_HW_HAS_SILABS_DCDC)) {
        sl_device_init_dcdc();
    }
```

gmarull commented on dts/bindings/regulator/silabs,series2-dcdc.yaml:
```diff
@@ -0,0 +1,63 @@
+    enum: [0, 1]
+
+  regulator-init-microvolt:
+    description: |
+      Output voltage for boost mode. Not used in buck mode.
```
no descriptions in standard properties please

gmarull commented on dts/bindings/regulator/silabs,series2-dcdc.yaml:
```diff
@@ -0,0 +1,63 @@
+      - 2300000
+      - 2400000
+
+  silabs,pfmx-peak-current-microamp:
```
why not milliamps? all values have a x1000 scale factor

asmellby replied:
I used microamp for symmetry with the standard regulator properties `regulator-*-microamp`. I agree that the scale factor of 1000 is silly, and if milliamp is OK, I will change to it.

gmarull commented on soc/silabs/silabs_s2/Kconfig:
```diff
@@ -10,4 +10,7 @@ config SOC_FAMILY_SILABS_S2
 
 rsource "*/Kconfig"
 
+config HW_HAS_SILABS_DCDC
+	def_bool $(dt_compat_enabled,$(DT_COMPAT_SILABS_SERIES2_DCDC))
```
please avoid these redundant Kconfig options and just use DT natively in C code

asmellby replied:
I had been pointed to https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nordic/common/Kconfig.peripherals in the past, but if this is a pattern that we don't want to propagate, and just use DT directly instead, that is completely fine by me :)

gmarull commented on soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h:
```diff
@@ -0,0 +1,65 @@
+#define SL_DEVICE_INIT_DCDC_CONFIG_H
+
+#include <zephyr/devicetree.h>
+#include <soc.h>
```
no soc.h please

gmarull commented on soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h:
```diff
@@ -0,0 +1,65 @@
+
+#define SL_DEVICE_INIT_DCDC_TYPE DT_PROP_OR(DCDC_NODE, regulator_initial_mode, 0)
+
+#define SL_DEVICE_INIT_DCDC_BOOST_OUTPUT                                                           \
+	(DT_ENUM_HAS_VALUE(DCDC_NODE, regulator_init_microvolt, 1800000)                           \
+		 ? emuDcdcBoostOutputVoltage_1v8                                                   \
+	 : DT_ENUM_HAS_VALUE(DCDC_NODE, regulator_init_microvolt, 1900000)                         \
+		 ? emuDcdcBoostOutputVoltage_1v9                                                   \
+	 : DT_ENUM_HAS_VALUE(DCDC_NODE, regulator_init_microvolt, 2000000)                         \
+		 ? emuDcdcBoostOutputVoltage_2v0                                                   \
+	 : DT_ENUM_HAS_VALUE(DCDC_NODE, regulator_init_microvolt, 2100000)                         \
+		 ? emuDcdcBoostOutputVoltage_2v1                                                   \
+	 : DT_ENUM_HAS_VALUE(DCDC_NODE, regulator_init_microvolt, 2200000)                         \
+		 ? emuDcdcBoostOutputVoltage_2v2                                                   \
+	 : DT_ENUM_HAS_VALUE(DCDC_NODE, regulator_init_microvolt, 2300000)                         \
+		 ? emuDcdcBoostOutputVoltage_2v3                                                   \
+	 : DT_ENUM_HAS_VALUE(DCDC_NODE, regulator_init_microvolt, 2400000)                         \
+		 ? emuDcdcBoostOutputVoltage_2v4                                                   \
+		 : emuDcdcBoostOutputVoltage_1v8)
```
aren't enums correlated to voltages so that something like linear range can be used? This looks seriously ugly.

gmarull commented on soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h:
```diff
@@ -0,0 +1,65 @@
+	DT_NODE_HAS_PROP(DCDC_NODE, silabs_pfmx_peak_current_microamp)
+
+#define SL_DEVICE_INIT_DCDC_PFMX_IPKVAL                                                            \
+	(DT_ENUM_HAS_VALUE(DCDC_NODE, silabs_pfmx_peak_current_microamp, 50000)                    \
+		 ? DCDC_PFMXCTRL_IPKVAL_LOAD50MA                                                   \
+	 : DT_ENUM_HAS_VALUE(DCDC_NODE, silabs_pfmx_peak_current_microamp, 65000)                  \
+		 ? DCDC_PFMXCTRL_IPKVAL_LOAD65MA                                                   \
+	 : DT_ENUM_HAS_VALUE(DCDC_NODE, silabs_pfmx_peak_current_microamp, 73000)                  \
```
same

gmarull commented on boards/silabs/radio_boards/slwrb4180a/slwrb4180a_defconfig:
```diff
@@ -5,6 +5,5 @@ CONFIG_CONSOLE=y
 CONFIG_GPIO=y
-CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=38400000
-CONFIG_CMU_HFCLK_HFXO=y
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=76800000
```
this setting should be grabbed from DT, see other examples in tree

gmarull commented on dts/arm/silabs/efr32mg21.dtsi:
```diff
@@ -17,6 +18,16 @@
+	clocks {
+		clk_hfxo: clk-hfxo {
+			#clock-cells = <0>;
+			compatible = "silabs,hfxo";
```
is this device addressable? if so, it should have a `@reg`.

asmellby committed:
soc: silabs: Introduce family specific defconfig

Defconfig was only available at the vendor and series level,
make it possible to have family specific definitions too.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

asmellby committed:
soc: silabs: Derive SYS_CLOCK_HW_CYCLES_PER_SEC from DT

On Series 2, set the SYS_CLOCK_HW_CYCLES_PER_SEC Kconfig option from
DeviceTree, rather than separately configuring it in board-level
defconfig.

Signed-off-by: Aksel Skauge Mellbye <aksel.mellbye@silabs.com>

jerome-pouiller commented on dts/bindings/clock/silabs,series2-lfrco.yaml:
```diff
@@ -0,0 +1,13 @@
+compatible: "silabs,series2-lfrco"
+
+description: |
+  Silicon Labs LFRCO peripheral.
```
Help messages are the right place to expand acronyms (in many case it provide a sufficient documentation).

jerome-pouiller commented on dts/bindings/clock/silabs,series2-lfxo.yaml:
```diff
@@ -0,0 +1,29 @@
+  mode:
+    type: string
+    description: |
+      HFXO mode. Defaults to "xtal", expecting a 32768 Hz crystal oscillator on XI and XO pins.
```
HFXO? (BTW, you can probably get rid of this description. "LFXO mode" is not very helpful when the attribute is named "mode" and the device "LFXO").

jerome-pouiller reviewed (dismissed):
I still have some concerns with use of `config/sl_*_config.h`. I believe we should work on a long term fix for these files. However, we can probably live with that for now.

That's said, I am happy to see much ugly code disappeared with this PR :-).

jhedberg commented:
@gmarull could you please take another look? Thanks.

jhedberg commented on soc/silabs/silabs_s2/config/sl_device_init_dcdc_config.h:
```diff
@@ -0,0 +1,35 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * This configuration header is used by the HAL driver device_init_dcdc from hal_silabs,
+ * invoked through the soc_early_init hook. DeviceTree options are converted to config macros
+ * expected by the HAL driver.
+ */
```
> shouldn;t this code be part of modules/hal_silabs? All glue code should reside in modules/

I'll let @asmellby comment on this specific SoC change, but in general we have a separate task of moving glue code out from the hal_silabs tree into `modules/hal_silabs` in the main tree (note: this directory doesn't even exist yet in the main tree). This will also include moving all Kconfig and CMake files from the `hal_silabs` tree to `modules/hal_silabs` in the main tree (i.e. we'll be adding `kconfig-ext: True` and `cmake-ext: True` to our `zephyr/module.yml`). Anyway that kind of refactoring is unrelated and out of scope for this PR.

asmellby replied:
Agree that modules/ may be the better place, and that should be handled by the later refactor. I added these files in soc/silabs/ since there already was an existing config file like this there (sl_device_init_hfxo_config.h).

asmellby commented on dts/arm/silabs/efr32bg2x.dtsi:
```diff
@@ -18,12 +18,133 @@
+			#clock-cells = <0>;
+			compatible = "fixed-factor-clock";
+			clocks = <&sysclk>;
+			/* Divisors 1, 2, 4, 8, 16 allowed */
```
Build will fail if invalid divisors are used. Agree that the comment may be redundant, and that separate DT bindings for every clock tree node declaring the valid enum may be the "technically correct" approach. However, it's very heavy handed, and would introduce another 20 or so bindings for Silabs. I can't see that any vendor has gone down this path before us in Zephyr. So I went for the comment as a light-weight documentation hint, as I found it better than just having the build fail.

gmarull reviewed (changes requested):
HAL glue code must go to `modules/hal_...`.

jhedberg commented:
> HAL glue code must go to `modules/hal_...`.

@gmarull it would be more helpful and productive if you engaged in the in the actual discussion of this topic rather than just repeating yourself: https://github.com/zephyrproject-rtos/zephyr/pull/79415#discussion_r1799346670

gmarull commented:
> > HAL glue code must go to `modules/hal_...`.
> 
> @gmarull it would be more helpful and productive if you engaged in the in the actual discussion thread for this topic rather than just repeating yourself: [#79415 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/79415#discussion_r1799346670)

This is new code, can perfectly be added to `modules/hal_silabs` right now. Code not related to this PR can of course be handled later.

jhedberg commented:
> This is new code, can perfectly be added to `modules/hal_silabs` right now. Code not related to this PR can of course be handled later.

@gmarull doesn't picking up a CMakeLists.txt from there require that the corresponding module has `cmake-ext: True`?(something that's part of the other refactoring I've been referring to and what we'd really like to get for Zephyr 4.0 (however starting that works depends on this PR getting merged first).

asmellby commented:
@gmarull Is my latest change what you had in mind? CMakeLists.txt will be updated to avoid referring to `modules/hal_silabs` from inside `soc/` in the follow-on PR that introduces `cmake-ext` for hal_silabs.

jhedberg committed:
modules: hal_silabs: Move CMake and Kconfig to the main tree

Move the CMakeLists.txt and Kconfig files from the hal_silabs tree under
modules/hal_silabs in the main tree. This also includes all Zephyr shim
code from the hal_silabs tree.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

asmellby commented:
Rebased on top of latest main to resolve conflict.

asmellby commented:
Rebased on top of @jhedberg's change in https://github.com/zephyrproject-rtos/zephyr/pull/80148, adding DNM label until his PR is merged.

rettichschnidi reviewed (commented):
Still works for the board `sim3u1xx_dk` 👍
